compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt

```diff
@@ -602,43 +603,61 @@ class FunctionInlining(
+                    substituteParameterViaTemporaryVariable()
                     return@forEach
                 }
 
-                // Arguments may reference the previous ones - substitute them.
-                val variableInitializer = argument.argumentExpression.transform(substitutor, data = null)
+                if (argument.isImmutableVariableLoad) {
+                    substituteMap[parameter] = variableInitializer
+                    return@forEach
+                }
 
                 val argumentExtracted = !argument.argumentExpression.isPure(false, context = context)
-
                 if (!argumentExtracted) {
-                    substituteMap[argument.parameter] = variableInitializer
+                    substituteMap[parameter] = variableInitializer
                 } else {
-                    val newVariable =
-                        currentScope.scope.createTemporaryVariable(
-                            irExpression = IrBlockImpl(
-                                variableInitializer.startOffset,
-                                variableInitializer.endOffset,
-                                variableInitializer.type,
-                                InlinerExpressionLocationHint((currentScope.irElement as IrSymbolOwner).symbol)
-                            ).apply {
-                                statements.add(variableInitializer)
-                            },
-                            nameHint = callee.symbol.owner.name.toString(),
-                            isMutable = false
-                        )
-
-                    evaluationStatements.add(newVariable)
-                    substituteMap[argument.parameter] = IrGetValueWithoutLocation(newVariable.symbol)
+                    substituteParameterViaTemporaryVariable()
                 }
             }
             return evaluationStatements
         }
+
+        private fun createTemporaryVariable(parameter: IrValueParameter, variableInitializer: IrExpression, callee: IrFunction): IrVariable {
+            val variable = currentScope.scope.createTemporaryVariable(
+                irExpression = IrBlockImpl(
+                    variableInitializer.startOffset,
+                    variableInitializer.endOffset,
+                    variableInitializer.type,
+                    InlinerExpressionLocationHint((currentScope.irElement as IrSymbolOwner).symbol)
+                ).apply {
+                    statements.add(variableInitializer)
+                },
+                nameHint = callee.symbol.owner.name.toString(),
+                isMutable = false,
+                origin = if (parameter == callee.extensionReceiverParameter) {
+                    IrDeclarationOrigin.IR_TEMPORARY_VARIABLE_FOR_INLINED_EXTENSION_RECEIVER
+                } else {
+                    IrDeclarationOrigin.IR_TEMPORARY_VARIABLE_FOR_INLINED_PARAMETER
+                }
+            )
+
+            if (alwaysCreateTemporaryVariablesForArguments) {
```

Comment:
I'm not sure if we should name the temporary variables like arguments only if the `alwaysCreateTemporaryVariablesForArguments` option is enabled or in all cases.